semgrep-core/src/core/ast/Ugly_print_AST.ml

```diff
@@ -0,0 +1,122 @@
+(* Nat Mote
+ *
+ * Copyright (C) 2019-2022 r2c
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public License
+ * version 2.1 as published by the Free Software Foundation, with the
+ * special exception on linking described in file LICENSE.
+ *
+ * This library is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the file
+ * LICENSE for more details.
+ *)
+
+open Common
+module G = AST_generic
+
+(******************************************************************************)
+(* UglyPrinter prints syntactically-correct code without attempting to format it
+ * nicely (as opposed to a pretty printer such as those based on Wadler's "A
+ * prettier printer" [1]). It would be difficult, if not impossible, to print
+ * code that conforms to everyone's style guides, linters, and formatters, so
+ * the prettifying of code is left to other tools.
+ *
+ * UglyPrinter can print code snippets, not just full programs. Any of the
+ * `print_` methods can be used by consumers.
+ *
+ * If UglyPrinter encounters a construct that it is unable to handle, it returns
+ * `None`. It is crucial that maintainers preserve this property, since
```

Comment:
why not raising an exception instead?

Reply:
I prefer to use `option` because callers are forced to handle it and cannot forget, especially for something like this where there isn't a clear single entry point where we could catch the exception and keep it internal to the module. If you feel strongly I can change it, though.